### PR TITLE
@types/sinon - add type inference for stub/spy arguments and return types

### PR DESCRIPTION
**Pull request by johnjesse**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.) - _No tests yet, wanted to discuss the change first_
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present). - _ran tsc and got no problems... it failed running tslint, it failed to assert the path_ `DefinitelyTyped/types/xxx` - _even though I have that folder structure... potentially a casing issue?_

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>> - _The documentation hasn't changed, my change just suggests that the typings can be tighter._
- [x] Increase the version number in the header if appropriate. - _No version change_
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

I've added generic type inference to the spy and stub calls. the key line for the spy is this:

```
    interface SinonSpyStatic {
        /**
         * Creates an anonymous function that records arguments, this value, exceptions and return values for all calls.
         */
        (): SinonSpy;
        /**
         * Spies on the provided function
         */
        (func: Function): SinonSpy;
        /**
         * Creates a spy for object.method and replaces the original method with the spy.
         * An exception is thrown if the property is not already a function.
         * The spy acts exactly like the original method in all cases.
         * The original method can be restored by calling object.method.restore().
         * The returned spy is the function object which replaced the original method. spy === object.method.
         */
        <T, K extends keyof T>(obj: T, method: K): T[K] extends (
            ...args: infer U
        ) => infer V
            ? SinonSpy<U, V>
            : SinonSpy;
    }
```

Basically, we can infer the types of the arguments of the function being stubbed and the return type of that function. Then we can use this within the spy/stub API such that, when you are accessing the arguments/return of a stub/spy you get type safety.

I haven't added tests yet, wanted to make sure people thought it was a good idea before I bothered. If we do think it's a good idea then I'll add them

### CODE REVIEW

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]

**Review by JoshuaKGoldberg — APPROVED** [review]

**Comment by johnjesse** [comment]
Does the build failure mean that everything that has `import Sinon = require("sinon");` needs to be referencing typescript 3? is there a quick way to find them all, or does the build highlight all the errors at once?

**Comment by PranavSenthilnathan** [comment]
Looks good to me. Let me know if you still wanted to write some tests. Otherwise I can merge it in now.

**Comment by johnjesse** [comment]
I've had a look and am ok not adding tests, I don't they'd add very much at the moment, Thanks

**Comment by PranavSenthilnathan** [comment]
Congratulations on your first DefinitelyTyped contribution! Thank you for being a part of the community!

File: types/sinon/index.d.ts [hunk 1]
```diff
@@ -523,7 +524,7 @@ declare namespace Sinon {
          * Makes the stub call the provided @param func when invoked.
          * @param func
          */
-        callsFake(func: (...args: U) => V): SinonStub<U, V>;
+        callsFake(func: () => V): SinonStub<U, V>;
```
Comment (egilsster) [inline types/sinon/index.d.ts]: @johnjesse Where is your source on this? The sinon source code, as far as I can tell (https://github.com/sinonjs/sinon/blob/master/lib/sinon/default-behaviors.js#L32-L34), allows a function with any amount of parameters to be passed in, simulating what would be called in the actual function. For example if you are stubbing an event emitter, you can have it call a fake function which takes in the entry being emitted as well as a call back:

```js
_emitter.expects('openReadStream').callsFake((_entry, callback) => {
  callback(null, readStream);
  readStream.emit('data', Buffer.from('Hello'));
  readStream.emit('data', Buffer.from('World'));
  readStream.emit('end');
});
```

Now I cant do this without some hacking around to by-pass the types. Clarification on this, or a revert, would be appreciated, thanks! :)

Comment (johnjesse) [reply types/sinon/index.d.ts]: I think you're right, I'll revert that commit and add a test case.

Comment (egilsster) [reply types/sinon/index.d.ts]: Aweome! Thank you so much, and many thanks for improving the sinon types! <3

Comment (johnjesse) [reply types/sinon/index.d.ts]: PR to fix is in, #30656

**Comment by tbrannam** [comment]
it seems that with the inferred addition to staticstubs - .withArgs() can no longer accept a SinonMatcher 

```
// works - uses default constructor
  const stub = sinon.stub()
  stub.withArgs(sinon.match.any).returns('test')

// doesn't pass typechecker - used new Inferred constructor
  const stub = sinon.stub(fs, 'readFileSync')
  stub.withArgs(sinon.match.any).returns('test')
```
